### PR TITLE
Defining interfaces for consuming/publishing changes

### PR DESCRIPTION
**Pull request by psasidhar**
Signed-off-by: Sasidhar Palaka <palakas@verizonmedia.com>

### CODE REVIEW

File: libs/java/server_common/src/main/java/com/yahoo/athenz/common/messaging/ChangeSubscriber.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.yahoo.athenz.common.messaging;
+
+import java.util.function.Consumer;
+
+/**
+ * Domain change subscriber interface
```
Comment (abvaidya) [inline libs/java/server_common/src/main/java/com/yahoo/athenz/common/messaging/ChangeSubscriber.java:22]: Domain is not necessary here

File: libs/java/server_common/src/main/java/com/yahoo/athenz/common/messaging/ChangeSubscriberFactory.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * limitations under the License.
+ */
+
+package com.yahoo.athenz.common.messaging;
+
+import com.yahoo.athenz.auth.PrivateKeyStore;
+
+public interface ChangeSubscriberFactory<T> {
+    /**
+     * creates a DomainChangeSubscriber
```
Comment (abvaidya) [inline libs/java/server_common/src/main/java/com/yahoo/athenz/common/messaging/ChangeSubscriberFactory.java:23]: Domain is not necessary here

File: libs/java/server_common/src/main/java/com/yahoo/athenz/common/messaging/ChangeSubscriberFactory.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+     * @param keyStore private keystore object for fetching any secrets if needed
+     * @param topicName of the topic to subscribe to
+     * @param SubscriptionName of the individual subscriber
```
Comment (abvaidya) [inline libs/java/server_common/src/main/java/com/yahoo/athenz/common/messaging/ChangeSubscriberFactory.java:26]: SubscriptionName - first char S should be in lower case

File: libs/java/server_common/src/main/java/com/yahoo/athenz/common/messaging/ChangeSubscriberFactory.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+     * @param type subscription mode as recognized by the underlying pub/sub system
+     * @return DomainChangeSubscriber that was just created
```
Comment (abvaidya) [inline libs/java/server_common/src/main/java/com/yahoo/athenz/common/messaging/ChangeSubscriberFactory.java:28]: Domain is not necessary here

File: libs/java/server_common/src/main/java/com/yahoo/athenz/common/messaging/ChangeSubscriberFactory.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+     */
+    default ChangeSubscriber<T> create(PrivateKeyStore keyStore, String topicName, String SubscriptionName, String type) {
```
Comment (abvaidya) [inline libs/java/server_common/src/main/java/com/yahoo/athenz/common/messaging/ChangeSubscriberFactory.java:30]: can we rename type to subscriptionType or something to make it explicit?

**Review by abvaidya — CHANGES_REQUESTED** [review]

File: libs/java/server_common/src/main/java/com/yahoo/athenz/common/messaging/ChangeSubscriberFactory.java (hunk 3 above)
Comment (psasidhar) [reply libs/java/server_common/src/main/java/com/yahoo/athenz/common/messaging/ChangeSubscriberFactory.java]: fixed the typo

File: libs/java/server_common/src/main/java/com/yahoo/athenz/common/messaging/ChangeSubscriberFactory.java (hunk 5 above)
Comment (psasidhar) [reply libs/java/server_common/src/main/java/com/yahoo/athenz/common/messaging/ChangeSubscriberFactory.java]: done

**Review by abvaidya — COMMENTED** [review]
Do we need default interface methods if we are not providing any reference / no-op implementations?

**Review by abvaidya — APPROVED** [review]
LGTM. @havetisyan could you please review?

File: libs/java/server_common/src/main/java/com/yahoo/athenz/common/messaging/ChangePublisherFactory.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+package com.yahoo.athenz.common.messaging;
+
+import com.yahoo.athenz.auth.PrivateKeyStore;
+
+public interface ChangePublisherFactory<T> {
+    /**
+     * creates a ChangePublisher
+     * @param keyStore private keystore object for fetching any secrets if needed
```
Comment (dvirguttman) [inline libs/java/server_common/src/main/java/com/yahoo/athenz/common/messaging/ChangePublisherFactory.java:24]: Why do we need it?
I think it should be implementation specific, and not part of the interface

**Review by havetisyan — APPROVED** [review]

**Pull request merged**